src/java.base/share/classes/jdk/internal/foreign/abi/s390x/sysv/SysVS390xVaList.java

```diff
@@ -33,123 +33,563 @@
+		Object argument = null;
+
+		switch (typeClass) {
+			case STRUCT_ONE_FLOAT -> {
+				long struArgSize = layout.byteSize();
+				long rightShiftBytes = isOverflowArea ? (VA_LIST_SLOT_BYTES - struArgSize) : 0;
+				MemorySegment struArgSegment = allocator.allocate(VA_LIST_SLOT_BYTES);
+				argument = struArgSegment.copyFrom(argAreaSegment.asSlice(rightShiftBytes, struArgSize));
+			}
+			case STRUCT -> {
+				/* There are two cases in handling struct arguments in the general register save
+				 * area or the overflow argument area:
+				 * 1)construct the struct with its address stored in a 8-byte stack slot of the
+				 *   memory area when the struct size is 3, 5, 6, 7 bytes or greater than 8 bytes.
+				 * 2)obtain all elements of the struct by copying them to the specified location
+				 *   when the struct's size is 1, 2, 4, 8 bytes.
+				 */
+				GroupLayout struLayout = (GroupLayout)layout;
+				if (isStruAddrRequired(struLayout)) {
+					long struArgSize = getAlignedStructSize(struLayout);
+					MemoryAddress struAddr = (MemoryAddress)argHandle.get(argAreaSegment);
+					MemorySegment struArgSegment = MemorySegment.ofAddress(struAddr, struArgSize, argAreaSegment.session());
+					argument = allocator.allocate(struArgSize).copyFrom(struArgSegment);
+				} else {
+					long struArgSize = struLayout.byteSize();
+					MemorySegment struArgSegment = allocator.allocate(VA_LIST_SLOT_BYTES);
+					argument = struArgSegment.copyFrom(argAreaSegment.asSlice(VA_LIST_SLOT_BYTES - struArgSize, struArgSize));
+				}
+			}
+			case INTEGER, POINTER, FLOAT -> {
+				/* A primitive/pointer is stored in a 8-byte stack slot of the memory area. */
+				argument = argHandle.get(argAreaSegment);
+			}
+			default -> throw new IllegalStateException("Unsupported TypeClass: " + typeClass);
+		}
+
+		return argument;
+	}
+
+	/* Check whether to store or obtain the struct's address in the memory area,
+	 * depending upon the size of struct.
+	 *
+	 * Note:
+	 * According to the ABI document, the struct's address is required only when
+	 * the struct's size is 3, 5, 6, 7 bytes or greater than 8 bytes; otherwise,
+	 * the elements of struct are copied to the memory area.
+	 */
+	private static boolean isStruAddrRequired(GroupLayout structLayout) {
+		long struArgSize = structLayout.byteSize();
+		boolean required = false;
+
+		if ((struArgSize != STRUCT_ARG_SIZE_1_BYTE)
+			&& (struArgSize != STRUCT_ARG_SIZE_2_BYTES)
+			&& (struArgSize != STRUCT_ARG_SIZE_4_BYTES)
+			&& (struArgSize != STRUCT_ARG_SIZE_8_BYTES)
+		) {
+			required = true;
+		}
+
+		return required;
+	}
+
+	/* Only a struct with its address stored in the slot needs to be aligned by 8 bytes to
+	 * ensure the elements of structs are correctly copied to the specified memory area.
+	 */
+	private static long getAlignedStructSize(GroupLayout structLayout) {
+		long struArgSize = structLayout.byteSize();
+
+		if ((struArgSize % VA_LIST_SLOT_BYTES) > 0) {
```

Comment:
I believe `struArgSize % VA_LIST_SLOT_BYTES` cannot be negative. If so it can be replaced by ` ((struArgSize % VA_LIST_SLOT_BYTES) != 0)`.

Reply:
Updated as suggested above.